Userland/Libraries/LibCore/Stream.h

```diff
@@ -0,0 +1,846 @@
+
+    bool is_eof() const { return !is_open() || m_last_read_was_eof; }
+    bool is_open() const { return m_fd != -1; }
+    void close();
+
+    KResultOr<size_t> pending_bytes() const;
+    KResultOr<bool> can_read_without_blocking(int timeout) const;
+
+    KResult set_blocking(bool enabled);
+    KResult set_close_on_exec(bool enabled);
+
+    void setup_notifier();
+    RefPtr<Core::Notifier> notifier() { return m_notifier; }
+
+private:
+    int m_fd { -1 };
+    bool m_last_read_was_eof { false };
+    RefPtr<Core::Notifier> m_notifier;
+};
+
+class TCPSocket final : public Socket {
+public:
+    static Result<TCPSocket, SocketError> connect(String const& host, u16 port);
+    static KResultOr<TCPSocket> connect(SocketAddress const& address);
+    static KResultOr<TCPSocket> adopt_fd(int fd);
+
+    TCPSocket(TCPSocket&& other)
+        : Socket(static_cast<Socket&&>(other))
+        , m_helper(move(other.m_helper))
+    {
+        if (is_open())
+            setup_notifier();
+    }
+
+    TCPSocket& operator=(TCPSocket&& other)
+    {
+        Socket::operator=(static_cast<Socket&&>(other));
+        m_helper = move(other.m_helper);
+        if (is_open())
+            setup_notifier();
+
+        return *this;
+    }
+
+    virtual bool is_readable() const override { return is_open(); }
+    virtual bool is_writable() const override { return is_open(); }
+    virtual KResultOr<size_t> read(Bytes buffer) override { return m_helper.read(buffer); }
+    virtual KResultOr<size_t> write(ReadonlyBytes buffer) override { return m_helper.write(buffer); }
+    virtual bool is_eof() const override { return m_helper.is_eof(); }
+    virtual bool is_open() const override { return m_helper.is_open(); };
+    virtual void close() override { m_helper.close(); };
+    virtual KResultOr<size_t> pending_bytes() const override { return m_helper.pending_bytes(); }
+    virtual KResultOr<bool> can_read_without_blocking(int timeout = 0) const override { return m_helper.can_read_without_blocking(timeout); }
+    KResult set_blocking(bool enabled) override { return m_helper.set_blocking(enabled); }
+    KResult set_close_on_exec(bool enabled) override { return m_helper.set_close_on_exec(enabled); }
+
+    virtual ~TCPSocket() override { close(); }
+
+private:
+    TCPSocket()
+    {
+    }
+
+    void setup_notifier()
+    {
+        VERIFY(is_open());
+
+        m_helper.setup_notifier();
+        m_helper.notifier()->on_ready_to_read = [this] {
+            if (on_ready_to_read)
+                on_ready_to_read();
+        };
+    }
+
+    PosixSocketHelper m_helper { Badge<TCPSocket> {} };
+};
+
+class UDPSocket final : public Socket {
+public:
+    static Result<UDPSocket, SocketError> connect(String const& host, u16 port);
+    static KResultOr<UDPSocket> connect(SocketAddress const& address);
+
+    UDPSocket(UDPSocket&& other)
+        : Socket(static_cast<Socket&&>(other))
+        , m_helper(move(other.m_helper))
```

Comment:
This is a bit weird, considering that the object is partially moved from.
it seems fine to me, but static analysers might flag this.

Reply:
I don't quite understand this, what do you mean by "object is partially moved from"? The default move constructor does the same thing as this, I believe.

Reply:
a slice of the object (the `Socket`-related bytes) are now in an unspecified state as they were moved from in `Socket(...)`, so the object is briefly in a partially moved-from state between `Socket(...)` and `m_helper(...)`.

That's not a problem, and I can't really see a way around it; this is more of an fyi.

Reply:
I see, thanks. As you said it shouldn't be a problem here.